src/mame/casio/zoomer.cpp

```diff
@@ -0,0 +1,728 @@
+}
+
+//**************************************************************************
+u32 zoomer_state::bank_base_a000(unsigned bank) const
+{
+	return (m_bank_a000[bank] & 0x7ff) << 14;
+}
+
+//**************************************************************************
+u8 zoomer_state::ems_2000_bank_r(offs_t offset)
+{
+	return m_bank_2000[offset >> 1] >> (BIT(offset, 0) * 8);;
+}
+
+//**************************************************************************
+void zoomer_state::ems_2000_bank_w(offs_t offset, u8 data)
+{
+	if (BIT(offset, 0))
+	{
+		m_bank_2000[offset >> 1] &= 0x00ff;
+		m_bank_2000[offset >> 1] |= (data << 8);
+	}
+	else
+	{
+		m_bank_2000[offset >> 1] &= 0xff00;
+		m_bank_2000[offset >> 1] |= data;
+	}
+}
+
+//**************************************************************************
+u8 zoomer_state::ems_a000_bank_r(offs_t offset)
+{
+	return m_bank_a000[offset >> 1] >> (BIT(offset, 0) * 8);
+}
+
+//**************************************************************************
+void zoomer_state::ems_a000_bank_w(offs_t offset, u8 data)
+{
+	if (BIT(offset, 0))
+	{
+		m_bank_a000[offset >> 1] &= 0x00ff;
+		m_bank_a000[offset >> 1] |= (data << 8);
+	}
+	else
+	{
+		m_bank_a000[offset >> 1] &= 0xff00;
+		m_bank_a000[offset >> 1] |= data;
+	}
+}
+
+//**************************************************************************
+u8 zoomer_state::ems_2000_r(offs_t offset)
+{
+	return ems_mem_r((offset & 0x1ffff) | bank_base_2000(offset >> 17));
+}
+
+//**************************************************************************
+void zoomer_state::ems_2000_w(offs_t offset, u8 data)
+{
+	ems_mem_w((offset & 0x1ffff) | bank_base_2000(offset >> 17), data);
+}
+
+//**************************************************************************
+u8 zoomer_state::ems_a000_r(offs_t offset)
+{
+	return ems_mem_r((offset & 0x3fff) | bank_base_a000(offset >> 14));
+}
+
+//**************************************************************************
+void zoomer_state::ems_a000_w(offs_t offset, u8 data)
+{
+	if (BIT(m_bank_a000[offset >> 14], 15))
+		irq_set<IRQ_EMS>(1); // write protect
+	else
+		ems_mem_w((offset & 0x3fff) | bank_base_a000(offset >> 14), data);
+}
+
+//**************************************************************************
+u8 zoomer_state::ems_mem_r(offs_t offset)
+{
+	if (offset < 0x0100000)
+		return m_nvram[offset];
+	else if (offset >= 0x0400000 && offset < 0x0800000)
+		return m_rom->base()[offset &= 0x3fffff];
+	// TODO: PCMCIA at 0x1000000
+	else if (!machine().side_effects_disabled())
+		logerror("%s: unmapped read from EMS @ %08x\n", machine().describe_context(), offset);
+
+	return 0;
+}
+
+//**************************************************************************
+void zoomer_state::ems_mem_w(offs_t offset, u8 data)
+{
+	if (offset < 0xf8000)
+		m_nvram[offset] = data;
+	// TODO: PCMCIA at 0x1000000
+	else
+		logerror("%s: unmapped write to EMS @ %08x\n", machine().describe_context(), offset);
+}
```

Comment:
As far as I can tell this should really use a combination of 2 `address_map_bank_device`s with shared address_map, and a `memory_view` that blocks writes where it matters (NVRAM only?)

Reply:
It'd presumably need a separate bank device for each window (128kb windows at 0x20000 and 16kb windows at 0xa0000) but I can do that if it still sounds reasonable.

Reply:
Does this look better?

Reply:
Definitely, thanks :)